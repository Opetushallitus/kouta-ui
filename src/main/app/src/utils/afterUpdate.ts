import { QueryClient, QueryKey } from 'react-query';

import { JULKAISUTILA } from '../constants';

export const afterUpdate = (
  queryClient: QueryClient,
  history,
  entityType: QueryKey,
  newTila: JULKAISUTILA
) => {
  const redirectToFrontpage = newTila && newTila === JULKAISUTILA.POISTETTU;
  const refetchActive = !newTila || newTila !== JULKAISUTILA.POISTETTU;

  queryClient.invalidateQueries(entityType, { refetchActive }).then(() => {
    if (redirectToFrontpage) {
      history.push('/');
    }
  });
};
