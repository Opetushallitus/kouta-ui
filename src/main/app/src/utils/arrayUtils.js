export const removeDuplicatesByFeature = (items, getFeatureValue) => Object.values(items.reduce((itemMap, item) => ({
    ...itemMap,
    [getFeatureValue(item)]: item
}), {}));

export const replaceItemsWithMatchingProperty = (items, replacementItem, propertyName) =>
    items.map(oldItem => oldItem[propertyName] === replacementItem[propertyName] ? replacementItem : oldItem);

