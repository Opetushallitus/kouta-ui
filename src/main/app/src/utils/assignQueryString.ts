import qs from 'query-string';

const assignQueryString = (queryString, object) => {
  const queryObj = qs.parse(queryString);

  return qs.stringify({ ...queryObj, ...object });
};

export default assignQueryString;
