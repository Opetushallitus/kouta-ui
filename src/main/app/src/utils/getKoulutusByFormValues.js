import { get, pick } from 'lodash';

const getKielivalinta = values => get(values, 'kieliversiot') || [];

const getKoulutusByFormValues = values => {
  const { muokkaaja, tila } = values;
  const kielivalinta = getKielivalinta(values);
  const tarjoajat = get(values, 'tarjoajat') || [];
  const koulutusKoodiUri = get(values, 'information.koulutus.value') || null;
  const koulutustyyppi = get(values, 'koulutustyyppi') || null;
  const osiot = get(values, 'lisatiedot.osiot') || [];
  const osioKuvaukset = get(values, 'lisatiedot.osioKuvaukset') || {};

  const osiotWithKuvaukset = osiot.map(({ value }) => ({
    otsikkoKoodiUri: value,
    teksti: pick(osioKuvaukset[value] || {}, kielivalinta),
  }));

  const kuvaus = pick(get(values, 'description.kuvaus') || {}, kielivalinta);
  const nimi = pick(get(values, 'information.nimi') || {}, kielivalinta);
  const opintojenLaajuusKoodiUri =
    get(values, 'information.opintojenLaajuus.value') || null;

  const tutkintonimikeKoodiUrit = (
    get(values, 'information.tutkintonimike') || []
  ).map(({ value }) => value);

  const koulutusalaKoodiUrit = (
    get(values, 'information.koulutusalat') || []
  ).map(({ value }) => value);

  const kuvauksenNimi = pick(
    get(values, 'description.nimi') || {},
    kielivalinta,
  );

  const teemakuva = get(values, 'teemakuva');

  const julkinen = Boolean(get(values, 'julkinen'));

  return {
    johtaaTutkintoon: true,
    muokkaaja,
    tila,
    kielivalinta,
    tarjoajat,
    koulutusKoodiUri,
    koulutustyyppi,
    nimi,
    julkinen,
    teemakuva,
    metadata: {
      tyyppi: koulutustyyppi,
      lisatiedot: osiotWithKuvaukset,
      kuvaus,
      opintojenLaajuusKoodiUri,
      tutkintonimikeKoodiUrit,
      kuvauksenNimi,
      koulutusalaKoodiUrit,
    },
  };
};

export default getKoulutusByFormValues;
