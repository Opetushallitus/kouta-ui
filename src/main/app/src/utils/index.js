import toPairs from 'lodash/toPairs';
import zipObject from 'lodash/zipObject';
import pick from 'lodash/pick';
import dateFnsformatDate from 'date-fns/format';
import padStart from 'lodash/padStart';
import memoizee from 'memoizee';
import flowRight from 'lodash/flowRight';

export const isString = value => typeof value === 'string';

export const isNumber = value => typeof value === 'number';

export const isDate = value => value instanceof Date;

export const isObject = value =>
  Object.prototype.toString.call(value) === '[object Object]';

export const isArray = value => Array.isArray(value);

export const isValidDate = value => isDate(value) && !isNaN(value);

export const formatDate = dateFnsformatDate;

export const isNumeric = value => {
  if (isNumber(value)) {
    return true;
  }

  if (isString(value)) {
    return !isNaN(parseFloat(value));
  }

  return false;
};

export const getLanguageValue = (value, language = 'fi') =>
  isObject(value) ? value[language] || null : null;

export const isFunction = value => typeof value === 'function';

export const getFirstLanguageValue = (value, priorityArg) => {
  const defaultPriority = ['fi', 'en', 'sv'];

  let priority = defaultPriority;

  if (isArray(priorityArg)) {
    priority = [...priorityArg, ...defaultPriority];
  }

  if (isString(priorityArg)) {
    priority = [priorityArg, ...defaultPriority];
  }

  // eslint-disable-next-line
  for (const p of priority) {
    const v = getLanguageValue(value, p);

    if (v) {
      return v;
    }
  }

  return null;
};

export const arrayToTranslationObject = (arr, languageField = 'kieli') => {
  return isArray(arr)
    ? arr.reduce((acc, curr) => {
        acc[
          isString(curr[languageField])
            ? curr[languageField].toLowerCase()
            : '_'
        ] = curr;

        return acc;
      }, {})
    : {};
};

export const getInvalidTranslations = (
  obj,
  languages = [],
  validate = v => !!v,
) => {
  if (!isObject(obj)) {
    return languages;
  }

  const translationObj = {
    ...zipObject(languages),
    ...pick(obj, languages),
  };

  const pairs = toPairs(translationObj);

  return pairs
    .filter(([, value]) => !validate(value))
    .map(([language]) => language);
};

export const getKoutaDateString = dateData => {
  if (isValidDate(dateData)) {
    return formatDate(dateData, `yyyy-MM-dd'T'HH:mm`);
  }

  if (!isObject(dateData)) {
    return null;
  }

  const { year, month, day, hour, minute } = dateData;

  return `${year}-${padStart(month, 2, '0')}-${padStart(day, 2, '0')}T${
    isNumeric(hour) ? padStart(hour, 2, '0') : '00'
  }:${isNumeric(minute) ? padStart(minute, 2, '0') : '00'}`;
};

export const formatKoutaDateString = (dateString, format) => {
  if (!isString(dateString)) {
    return '';
  }

  const [date, time = ''] = dateString.split('T');
  const [year, month, day] = date.split('-');
  const [hour = '0', minute = '0'] = time.split(':');

  let formattedDate = format;

  formattedDate = formattedDate.replace(/DD/g, padStart(day, 2, '0'));
  formattedDate = formattedDate.replace(/MM/g, padStart(month, 2, '0'));
  formattedDate = formattedDate.replace(/YYYY/g, year);
  formattedDate = formattedDate.replace(/HH/g, padStart(hour, 2, '0'));
  formattedDate = formattedDate.replace(/mm/g, padStart(minute, 2, '0'));

  return formattedDate;
};

export const createChainedFunction = (...fns) => (...args) => {
  // eslint-disable-next-line
  for (const fn of fns) {
    if (isFunction(fn)) {
      fn(...args);
    }
  }
};

export const memoize = (fn, opts = {}) => memoizee(fn, { max: 100, ...opts });

export const memoizePromise = (fn, opts = {}) =>
  memoize(fn, { promise: true, ...opts });

export const compose = flowRight;

export const noop = () => {};

export const isBoolean = value => typeof value === 'boolean';

export const getTestIdProps = testId => ({
  'data-test-id': testId,
});
