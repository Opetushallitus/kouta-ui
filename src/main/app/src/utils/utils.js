const connectionMap = {};
let listenerId = 0;


const parseJson = (jsonString) => {
  if (!jsonString) {
    return null;
  }
  try {
    return JSON.parse(jsonString);
  } catch(e) {
    return null;
  }
}

const getItem = (eventName) => parseJson(localStorage.getItem(eventName));

const enforceItem = (eventName) => getItem(eventName) || {};

const setItem = (eventName, item) => localStorage.setItem(eventName, JSON.stringify(item));

export const observe = (eventName, item) => setItem(eventName, clone(item));

export const updateState = (eventName, item) => {
  let targetItem = enforceItem(eventName);
  targetItem = Object.assign(targetItem, item);
  setItem(eventName, targetItem);
  broadcast(eventName, targetItem);
  return safeClone(targetItem);
}

export const containsState = (eventName, property) => getItem(eventName) !== null;

export const getState = (eventName, property) => {
  const state = getItem(eventName);
  return safeClone(property ? state[property] : state);
}

//retrieves ID from listener and creates one if it does not exist yet
const enforceListenerId = (listener) => {
  if (!listener._listenerId) {
    listenerId += 1;
    listener._listenerId = listenerId.toString();
  }
  return listener._listenerId;
}

// a simple function to couple a listener with matching event
export const connect = (eventName, listener, targetFunction) => {
  const listenerId = enforceListenerId(listener);
  connectionMap[eventName] = connectionMap[eventName] || {};
  connectionMap[eventName][listenerId.toString()] = targetFunction;
  const dataItem = getItem(eventName);
  if (dataItem) {
    broadcast(eventName, dataItem);
  }
};

const clone = (item) => {
  try {
    return JSON.parse(JSON.stringify(item));
  } catch(e) {
    return item;
  }
}

const safeClone = (item) => item ? clone(item) : null;

//A simple callback mechanism to achieve what MobX can't do: 1) pass change when value is changed inside a nested object, and 2) work with inheritance
export const broadcast = (eventName, item) => Object.values(connectionMap[eventName] || {}).forEach((listener) => listener(clone(item)));



