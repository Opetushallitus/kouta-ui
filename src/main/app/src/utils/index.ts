import dateFnsformatDate from 'date-fns/format';
import _fp from 'lodash/fp';
import _ from 'lodash';
import stripTags from 'striptags';
import { ALLOWED_HTML_TAGS, LANGUAGES } from '#/src/constants';
import { memoize } from '#/src/utils/memoize';
import {
  isEditorState,
  isEmptyEditorState,
} from '#/src/components/Editor/utils';

const { NODE_ENV, REACT_APP_CYPRESS } = process.env;

export const isDev = NODE_ENV === 'development';

export const isNodeEnv = env =>
  (_.isArray(env) ? env : [env]).includes(NODE_ENV);

export const isCypress = Boolean(REACT_APP_CYPRESS);

export const isValidDate = value => _.isDate(value) && !_.isNaN(value);

export const isPartialDate = date => {
  const dateString = _.isDate(date) ? getKoutaDateString(date) : null;
  if (dateString) {
    const [date] = dateString.split('T');
    return date === 'NaN-NaN-NaN';
  } else {
    return false;
  }
};

export const formatDate = dateFnsformatDate;

export const isNumeric = value => {
  if (_.isNumber(value)) {
    return true;
  }

  if (_.isString(value)) {
    return !_.isNaN(parseFloat(value));
  }

  return false;
};

export const getReadableDateTime = dateData => {
  try {
    return formatDate(dateData, 'd.M.yyyy HH:mm');
  } catch {
    return '-';
  }
};

export const getKoutaDateString = dateData => {
  if (isValidDate(dateData)) {
    return formatDate(dateData, `yyyy-MM-dd'T'HH:mm`);
  }

  if (!_.isObject(dateData)) {
    return null;
  }

  const { year, month, day, hour, minute } = dateData;

  return `${year}-${_.padStart(month, 2, '0')}-${_.padStart(day, 2, '0')}T${
    isNumeric(hour) ? _.padStart(hour, 2, '0') : '00'
  }:${isNumeric(minute) ? _.padStart(minute, 2, '0') : '00'}`;
};

export const formatKoutaDateString = (dateString, format) => {
  if (!_.isString(dateString)) {
    return '';
  }

  const [date, time = ''] = dateString.split('T');
  const [year, month, day] = date.split('-');
  const [hour = '0', minute = '0'] = time.split(':');

  let formattedDate = format;

  formattedDate = formattedDate.replace(/DD/g, _.padStart(day, 2, '0'));
  formattedDate = formattedDate.replace(/MM/g, _.padStart(month, 2, '0'));
  formattedDate = formattedDate.replace(/YYYY/g, year);
  formattedDate = formattedDate.replace(/HH/g, _.padStart(hour, 2, '0'));
  formattedDate = formattedDate.replace(/mm/g, _.padStart(minute, 2, '0'));

  return formattedDate;
};

export const createChainedFunction = (...fns) => (...args) => {
  // eslint-disable-next-line
  for (const fn of fns) {
    if (_.isFunction(fn)) {
      fn(...args);
    }
  }
};

export const getTestIdProps = testId => ({
  'data-testid': testId,
});

export const getImageFileDimensions = imgFile => {
  const objectURL = URL.createObjectURL(imgFile);
  const img = new Image();
  img.src = objectURL;
  const result = new Promise((resolve, reject) => {
    img.onload = () => resolve({ width: img.width, height: img.height });
    img.onerror = e => reject(e);
  });
  result.finally(() => URL.revokeObjectURL(objectURL));
  return result;
};

export const getFileExtension = file => {
  const parts = file.name.split('.');
  return parts.length > 1 ? _.last(parts).toLowerCase() : '';
};

/**
 * Check that given predicate returns truthy for a value or any value in an array
 * Can be used with lodash.cond() to improve readability
 * @param {*} value A single value or an array of values to check
 * @param {*} predicate A function returning truthy for a single matching value
 */
export const ifAny = value => predicate =>
  _.isArray(value) ? value.some(predicate) : predicate(value);

export const otherwise = () => true;

export const sanitizeHTML = html => stripTags(html, ALLOWED_HTML_TAGS);

export const parseKeyVal = memoize(
  _fp.compose(
    _fp.fromPairs,
    _fp.map(keyVal => keyVal.split('=')),
    _fp.split(';')
  )
);

export const getCookie = name => _.get(parseKeyVal(document.cookie), name);

const allFuncs = (...fns) => value => _.every(fns, fn => fn(value));

export const formValueExists = value =>
  _.cond([
    [_.isNil, _.stubFalse],
    [
      allFuncs(
        v => _.isArray(v) || _.isString(v),
        v => v.length === 0
      ),
      _.stubFalse,
    ],
    [allFuncs(_.isPlainObject, _.isEmpty), _.stubFalse],
    [allFuncs(_.isPlainObject, v => v.value === ''), _.stubFalse],
    [allFuncs(isEditorState, isEmptyEditorState), _.stubFalse],
    [otherwise, _.stubTrue],
  ])(value);

export const isDeepEmptyFormValues = value =>
  !formValueExists(value) ||
  (_.isObjectLike(value) && _.every(value, isDeepEmptyFormValues));

export const assert = console.assert;

export const oneAndOnlyOne = all => all && all.length === 1 && all[0];

/** Tries to parse a (form) value to a number.
 * For empty values (null, undefined or empty string) returns null and when conversion to number fails, returns the given value.
 * This way we can pass a number representation to backend when needed, and give null when the value should be removed,
 * but fall back to original value when conversion fails, which is less confusing when debugging.
 */
export const maybeParseNumber = value => {
  if (_.isNil(value) || value === '') {
    return null;
  }
  const numberValue = Number(value);
  return _.isNaN(numberValue) ? value : numberValue;
};

export const toSelectValue = value => (_.isNil(value) ? null : { value });

const getFieldName = name =>
  name.match(`(.+?)(\\.${LANGUAGES.join('|')})?$`)?.[1];

// Get form values for saving. Filters out fields that user has hidden.
// Result can be sassed to get**ByFormValues().
export const getValuesForSaving = (
  values,
  registeredFields,
  unregisteredFields,
  initialValues = {}
) => {
  // Use initial values as a base. Especially important for editing.
  const saveableValues: any = initialValues;

  // Ensure that all fields that were unregistered (hidden by the user) are sent to backend as empty values
  _.each(unregisteredFields, ({ name }) => {
    const fieldName = getFieldName(name);
    _.set(saveableValues, fieldName, null);
  });

  // Ensure that the fields that are registered (visible) will be saved
  _.each(registeredFields, ({ name }) => {
    const fieldName = getFieldName(name);
    _.set(saveableValues, fieldName, _.get(values, fieldName));
  });

  // Some exceptions (fields that should be saved even though they are not visible)
  // TODO: There might be a few other exceptions. Check before using in other forms and add here.
  _.set(saveableValues, 'koulutustyyppi', values?.koulutustyyppi);
  _.set(saveableValues, 'muokkaaja', values?.muokkaaja);
  return saveableValues;
};
