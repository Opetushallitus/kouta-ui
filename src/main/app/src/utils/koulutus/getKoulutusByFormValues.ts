import _fp from 'lodash/fp';

import { serializeEditorState } from '#/src/components/Editor/utils';
import {
  KOULUTUSTYYPPI,
  MaaraTyyppi,
  TUTKINTOON_JOHTAVAT_KOULUTUSTYYPIT,
} from '#/src/constants';
import {
  InformationSectionValues,
  KoulutusFormValues,
  TutkinnonOsa,
} from '#/src/types/koulutusTypes';
import { maybeParseNumber, parseFloatComma, valueToArray } from '#/src/utils';
import { isTutkintoonJohtavaKorkeakoulutus } from '#/src/utils/koulutus/isTutkintoonJohtavaKorkeakoulutus';

const osaamisalaKoodiToKoodiUri = value =>
  value ? `osaamisala_${value}` : null;

function getKoulutuksetKoodiUri(
  osaamisala,
  koulutustyyppi: KOULUTUSTYYPPI,
  information?: InformationSectionValues
): Array<string> {
  if (isTutkintoonJohtavaKorkeakoulutus(koulutustyyppi)) {
    return _fp.map(koodi => koodi.value, information?.korkeakoulutukset);
  }

  if (koulutustyyppi === KOULUTUSTYYPPI.OSAAMISALA) {
    return valueToArray(osaamisala?.koulutus?.value);
  }

  return valueToArray(information?.koulutus?.value);
}

const getKoulutusByFormValues = (values: KoulutusFormValues) => {
  const { muokkaaja, tila, esikatselu = false } = values;
  const kielivalinta = values?.kieliversiot ?? [];
  const pickTranslations = _fp.pick(kielivalinta);

  const pohjanTarjoajat = values?.pohja?.tarjoajat;
  const kaytaPohjanJarjestajaa =
    values?.tarjoajat?.kaytaPohjanJarjestajaa ?? false;

  const koulutustyyppi = values?.koulutustyyppi || null;
  const osiot = values?.lisatiedot?.osiot ?? [];
  const osaamisala = values?.osaamisala;

  const sorakuvausId = values?.soraKuvaus?.value || null;

  const isLaajuusRange =
    values?.information?.laajuusNumeroTyyppi === MaaraTyyppi.VAIHTELUVALI;

  return {
    organisaatioOid: values?.organisaatioOid?.value,
    externalId: _fp.isEmpty(values?.externalId) ? null : values?.externalId,
    johtaaTutkintoon:
      TUTKINTOON_JOHTAVAT_KOULUTUSTYYPIT.includes(koulutustyyppi),
    muokkaaja,
    tila,
    kielivalinta,
    tarjoajat:
      pohjanTarjoajat && kaytaPohjanJarjestajaa
        ? pohjanTarjoajat
        : values?.tarjoajat?.tarjoajat || [],
    koulutuksetKoodiUri: getKoulutuksetKoodiUri(
      osaamisala,
      koulutustyyppi,
      values?.information
    ),
    koulutustyyppi,
    nimi:
      koulutustyyppi === KOULUTUSTYYPPI.TUTKINNON_OSA
        ? pickTranslations(values?.tutkinnonosat?.nimi ?? {})
        : pickTranslations(values?.information?.nimi ?? {}),
    julkinen: Boolean(values?.julkinen),
    esikatselu,
    ePerusteId: maybeParseNumber(
      values?.information?.eperuste?.value || osaamisala?.eperuste?.value
    ),
    teemakuva: values?.teemakuva,
    sorakuvausId,
    metadata: {
      tyyppi: koulutustyyppi,
      tutkinnonOsat: _fp.reduce(
        (
          resultOsat,
          {
            eperuste: { value: ePerusteId },
            koulutus: { value: koulutusKoodiUri },
            osat,
          }: TutkinnonOsa
        ) => [
          ...resultOsat,
          ..._fp.map(({ value, viite }) => ({
            ePerusteId: maybeParseNumber(ePerusteId),
            koulutusKoodiUri,
            tutkinnonosaId: maybeParseNumber(value),
            tutkinnonosaViite: maybeParseNumber(viite),
          }))(osat),
        ],
        [] as Array<{
          ePerusteId: number;
          koulutusKoodiUri?: string;
          tutkinnonosaId: number;
          tutkinnonosaViite: number;
        }>
      )(values?.tutkinnonosat?.osat),
      osaamisalaKoodiUri: osaamisalaKoodiToKoodiUri(
        osaamisala?.osaamisala?.value
      ),
      koulutustyyppi,
      lisatiedot: osiot.map(({ value }) => ({
        otsikkoKoodiUri: value,
        teksti: _fp.flow(
          pickTranslations,
          _fp.mapValues(serializeEditorState)
        )(values?.lisatiedot?.osioKuvaukset?.[value] ?? {}),
      })),
      kuvaus: _fp.flow(
        pickTranslations,
        _fp.mapValues(serializeEditorState)
      )(values?.description?.kuvaus ?? {}),
      opintojenLaajuusKoodiUri:
        values?.information?.opintojenLaajuus?.value || null,
      opintojenLaajuusyksikkoKoodiUri:
        values?.information?.opintojenLaajuusyksikko?.value || null,
      opintojenLaajuusNumero:
        parseFloatComma(values.information.opintojenLaajuusNumero) || null,
      opintojenLaajuusNumeroMin:
        parseFloatComma(values?.information?.opintojenLaajuusNumeroMin) || null,
      opintojenLaajuusNumeroMax:
        parseFloatComma(
          isLaajuusRange
            ? maybeParseNumber(values?.information?.opintojenLaajuusNumeroMax)
            : maybeParseNumber(values?.information?.opintojenLaajuusNumeroMin)
        ) || null,
      tutkintonimikeKoodiUrit: (values?.information?.tutkintonimike ?? []).map(
        ({ value }) => value
      ),
      koulutusalaKoodiUrit: (values?.information?.koulutusalat ?? []).map(
        ({ value }) => value
      ),
      linkkiEPerusteisiin: pickTranslations(
        values?.description?.linkkiEPerusteisiin ?? {}
      ),
      avoinKorkeakoulutus: values?.information?.avoinKorkeakoulutus || false,
      tunniste: values?.information?.tunniste || null,
    },
  };
};

export default getKoulutusByFormValues;
