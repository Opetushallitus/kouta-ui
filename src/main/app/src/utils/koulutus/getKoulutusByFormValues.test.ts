import { parseEditorState } from '#/src/components/Editor/utils';
import getKoulutusByFormValues from '#/src/utils/koulutus/getKoulutusByFormValues';

test('getKoulutusByFormValues returns correct koulutus given form values', () => {
  const koulutus = getKoulutusByFormValues({
    tila: 'tallennettu',
    muokkaaja: '1.1.1.1',
    kieliversiot: ['fi', 'sv'],
    tarjoajat: {
      tarjoajat: ['123.456.789'],
    },
    information: {
      nimi: {
        fi: 'Fi nimi',
        sv: 'Sv nimi',
      },
      koulutus: {
        value: 'koulutuskoodi_1#1',
      },
      korkeakoulutukset: ['koulutus_371101#1', 'koulutus_201000#1'],
      opintojenLaajuus: {
        value: 'laajuus_1#1',
      },
      tutkintonimike: [{ value: 'nimike_1#1' }, { value: 'nimike_2#1' }],
      koulutusalat: [
        { value: 'koulutusala_1#1' },
        { value: 'koulutusala_2#1' },
      ],
    },
    koulutustyyppi: 'yo',
    lisatiedot: {
      osioKuvaukset: {
        'osio_1#1': {},
        'osio_2#1': {
          fi: parseEditorState('Fi kuvaus'),
          sv: parseEditorState('Sv kuvaus'),
        },
      },
      osiot: [{ value: 'osio_1#1' }, { value: 'osio_2#1' }],
    },
    description: {
      kuvaus: {
        fi: parseEditorState('Fi kuvaus'),
        sv: parseEditorState('Sv kuvaus'),
      },
      nimi: {
        fi: 'Fi nimi',
        sv: 'Sv nimi',
      },
    },
    tutkinnonosat: {
      osat: [
        {
          eperuste: { value: '1234' },
          koulutus: { value: 'koulutuskoodi_2#1' },
          tutkinnonosa: { value: '567567' },
          tutkinnonosaviite: '9847598475',
        },
      ],
    },
    julkinen: true,
  });

  expect(koulutus).toMatchSnapshot();
});
