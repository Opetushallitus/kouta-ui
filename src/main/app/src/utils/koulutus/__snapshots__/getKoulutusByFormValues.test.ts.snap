// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getKoulutusByFormValues returns correct koulutus given form values 1`] = `
Object {
  "ePerusteId": null,
  "esikatselu": undefined,
  "johtaaTutkintoon": true,
  "julkinen": true,
  "kielivalinta": Array [
    "fi",
    "sv",
  ],
  "koulutusKoodiUri": "koulutuskoodi_1#1",
  "koulutustyyppi": "yo",
  "metadata": Object {
    "koulutusalaKoodiUrit": Array [
      "koulutusala_1#1",
      "koulutusala_2#1",
    ],
    "kuvauksenNimi": Object {
      "fi": "Fi nimi",
      "sv": "Sv nimi",
    },
    "kuvaus": Object {
      "fi": "Fi kuvaus",
      "sv": "Sv kuvaus",
    },
    "lisatiedot": Array [
      Object {
        "otsikkoKoodiUri": "osio_1#1",
        "teksti": Object {},
      },
      Object {
        "otsikkoKoodiUri": "osio_2#1",
        "teksti": Object {
          "fi": "Fi kuvaus",
          "sv": "Sv kuvaus",
        },
      },
    ],
    "opintojenLaajuusKoodiUri": "laajuus_1#1",
    "tutkinnonOsat": Array [
      Object {
        "ePerusteId": 1234,
        "koulutusKoodiUri": "koulutuskoodi_2#1",
        "tutkinnonosaId": 567567,
        "tutkinnonosaViite": 9847598475,
      },
    ],
    "tutkintonimikeKoodiUrit": Array [
      "nimike_1#1",
      "nimike_2#1",
    ],
    "tyyppi": "yo",
  },
  "muokkaaja": "1.1.1.1",
  "nimi": Object {
    "fi": "Fi nimi",
    "sv": "Sv nimi",
  },
  "tarjoajat": Array [
    "123.456.789",
  ],
  "teemakuva": undefined,
  "tila": "tallennettu",
}
`;
